### PR TITLE
Refactor escape

### PR DESCRIPTION
**Pull request by thewilsonator**

### CODE REVIEW

File: src/dmd/escape.d [hunk 1]
```diff
@@ -154,71 +154,51 @@ bool checkMutableArguments(Scope* sc, FuncDeclaration fd, TypeFunction tf,
             escapeByValue(arg, &eb.er);
     }
 
-    foreach (const i, ref eb; escapeBy[0 .. $ - 1])
+    void checkOnePair(size_t i,ref EscapeBy eb, ref EscapeBy eb2,
```
Comment (MoonlightSentinel) [inline src/dmd/escape.d]: A small DDoc comment and maybe a more descriptive name would be nice

Comment (MoonlightSentinel) [reply src/dmd/escape.d]:
```suggestion
    void checkOnePair(size_t i, ref EscapeBy eb, ref EscapeBy eb2,
```

File: src/dmd/escape.d [hunk 2]
```diff
@@ -154,71 +154,52 @@ bool checkMutableArguments(Scope* sc, FuncDeclaration fd, TypeFunction tf,
...
-                                    fd ? fd.toPrettyChars() : "indirectly");
-                            }
-                            errors = true;
-                        }
-                    }
-                }
-            }
+            const(char)* msg = eb.isMutable && eb2.isMutable
+            ? "more than one mutable reference %s `%s` in arguments to `%s()`"
+            : "mutable and const references %s `%s` in arguments to `%s()`";
```
Comment (thewilsonator) [inline src/dmd/escape.d]: (ignoring the fact that this doesn't compile as is) I'm wondering why we give two different, but almost identical, error messages here. I for one am all for giving the same message.

File: src/dmd/escape.d [hunk 3]
```diff
@@ -154,71 +154,52 @@ bool checkMutableArguments(Scope* sc, FuncDeclaration fd, TypeFunction tf,
...
-                        }
-                    }
+                    checkOnePair(i, eb, eb2, v, v2);
                 }
             }
         }
     }
+    foreach (const i, ref eb; escapeBy[0 .. $ - 1])
+    {
+        escape!"byvalue"(i, eb, true);
```
Comment (Geod24) [inline src/dmd/escape.d]: Why the mixin ? You could just pass a `ref` to `byref` / `byvalue` no ?

File: src/dmd/escape.d [hunk 4]
```diff
@@ -789,26 +789,25 @@ ByRef:
         if (!(va && va.isScope()))
             notMaybeScope(v);
 
-        if ((v.storage_class & (STC.ref_ | STC.out_)) == 0 && p == sc.func)
+        if (!((v.storage_class & (STC.ref_ | STC.out_)) == 0 && p == sc.func))
```
Comment (Geod24) [inline src/dmd/escape.d]: While a valid refactor, double negatives can be expressed much more intelligibly:

```suggestion
        if ((v.storage_class & (STC.ref_ | STC.out_)) || p != sc.func)
```

File: src/dmd/escape.d [hunk 5]
```diff
@@ -835,25 +834,24 @@ ByRef:
             if (!(va && va.isScope()))
                 notMaybeScope(v);
 
-            if ((v.storage_class & (STC.ref_ | STC.out_ | STC.scope_)) && p == sc.func)
+            if (!((v.storage_class & (STC.ref_ | STC.out_ | STC.scope_)) && p == sc.func))
```
Comment (Geod24) [inline src/dmd/escape.d]:
```suggestion
            if ((v.storage_class & (STC.ref_ | STC.out_ | STC.scope_)) || p != sc.func)
```

File: src/dmd/escape.d [hunk 6]
```diff
@@ -1054,40 +1054,39 @@ bool checkNewEscape(Scope* sc, Expression e, bool gag)
         /* Check for returning a ref variable by 'ref', but should be 'return ref'
          * Infer the addition of 'return', or set result to be the offending expression.
          */
-        if (v.storage_class & (STC.ref_ | STC.out_))
+        if (!(v.storage_class & (STC.ref_ | STC.out_)))
+            continue;
+
+        // DIP25
```
Comment (Geod24) [inline src/dmd/escape.d]: That's not quite correct anymore. The comment DIP25 applied to the block under.
Now that I look at this, it does look a bit weird. I suppose the rationale for skipping DIP25 on non-root module is that we don't do semantic on functions body if we don't compile them (unless they're `auto`). But then we shouldn't be in this function to begin with.

File: src/dmd/escape.d [hunk 7]
```diff
@@ -1054,40 +1054,39 @@ bool checkNewEscape(Scope* sc, Expression e, bool gag)
...
+        if (!(sc._module && sc._module.isRoot()))
```
Comment (Geod24) [inline src/dmd/escape.d]:
```suggestion
        if (!sc._module || !sc._module.isRoot())
```

File: src/dmd/escape.d (hunk 2 above)
Comment (Geod24) [reply src/dmd/escape.d]: What message would that be ?

**Review by Geod24 — COMMENTED** [review]
Looks nice overall, a few comment.
If you want this split commit-by-commit, merge the "fix test cases" and "ws" commits with the commits they fix. I usually review commit by commit when possible so this threw me off a bit.

File: src/dmd/escape.d (hunk 2 above)
Comment (thewilsonator) [reply src/dmd/escape.d]: byval escaping ref print `reference to`. escaping  by ref prints `reference of`. It took me a while to see what the difference was when looking at the error message.

File: src/dmd/escape.d (hunk 3 above)
Comment (thewilsonator) [reply src/dmd/escape.d]: while eb is fine, eb2 in `escape` is not, it needs to know which field to access. I suppose I could do it with `.offsetof` or `.tupleof` but I'd rather use a mixin than those. If you have better idea please do tell.

File: src/dmd/escape.d (hunk 6 above)
Comment (thewilsonator) [reply src/dmd/escape.d]: Hmm, do you think the code comment below is fine on its own? If so I'll just remove this.

File: src/dmd/escape.d (hunk 5 above)
Comment (thewilsonator) [reply src/dmd/escape.d]: Ahh that was a misplaced `)` the whole expression should be negated.

File: src/dmd/escape.d (hunk 2 above)
Comment (WalterBright) [reply src/dmd/escape.d]: > I'm wondering why we give two different, but almost identical, error messages here.

I often make them slightly different as it makes it easier to find which one is tripping.

Comment (thewilsonator) [reply src/dmd/escape.d]: Fair enough.

File: src/dmd/escape.d (hunk 6 above)
Comment (WalterBright) [reply src/dmd/escape.d]: The code in escape.d is pretty sensitive, please make darn sure that a refactoring here does NOT introduce ANY behavioral changes.

Comment (thewilsonator) [reply src/dmd/escape.d]: This particular is just comment placement. Previously it was a comment followed by an `if` block. This PR changes the `if` to an early continue.

Point taken though.

File: src/dmd/escape.d (hunk 3 above)
Comment (WalterBright) [reply src/dmd/escape.d]:
```
foreach (VarDeclaration v2; by == byref ? eb2.er.byref : eb2.er.byvalue)
```

**Comment by thewilsonator** [comment]
Finally, its green!

**Comment by thewilsonator** [comment]
@Geod24

**Pull request merged**